show-device-location-using-indoor-positioning/src/main/java/com/esri/arcgismaps/sample/showdevicelocationusingindoorpositioning/MainActivity.kt

```diff
@@ -0,0 +1,361 @@
+                                // positioningID - an ID which identifies a specific row in the positioningTable that should be used for setting up IPS.
+                                mIndoorsLocationDataSource = IndoorsLocationDataSource(
+                                    positioningFeatureTable,
+                                    getPathwaysTable(),
+                                    getLevelsTable(),
+                                    positioningId
+                                )
+                                // start the location display (blue dot)
+                                startLocationDisplay()
+                            } else {
+                                showError("No features found")
+                            }
+                        }
+                    }?.onFailure {
+                        showError("The positioning table contain no data")
+                    }
+            }
+        } else {
+            showError("Positioning Table not found in FeatureTables")
+        }
+    }
+
+    /**
+     * Find the exact formatting of the name "DateCreated" in the list of ServiceFeatureTable fields.
+     */
+    private fun getDateCreatedFieldName(fields: List<Field>): String? {
+        val field = fields.find {
+            it.name.equals(
+                "DateCreated",
+                ignoreCase = true
+            ) || it.name.equals("Date_Created", ignoreCase = true)
+        }
+        return field?.name
+    }
+
+    /**
+     * Retrieves the PathwaysTable which is an [ArcGISFeatureTable] that contains pathways as per the ArcGIS Indoors Information Model.
+     * Setting this property enables path snapping of locations provided by the IndoorsLocationDataSource.
+     */
+    private fun getPathwaysTable(): ArcGISFeatureTable? {
+        return try {
+            val pathwaysFeatureLayer =
+                mapView.map?.operationalLayers?.firstOrNull { it.name == "Pathways" } as? FeatureLayer
+            pathwaysFeatureLayer?.featureTable as? ArcGISFeatureTable
+        } catch (e: Exception) {
+            // if pathways table not found in map's operationalLayers
+            showError("Pathways table not found")
+            null
+        }
+    }
+
+    /**
+     * Retrieves the Levels tables which is an [ArcGISFeatureTable] that contains floor levels in accordance with the ArcGIS Indoors Information Model.
+     * Providing this table enables the retrieval of a location's floor level ID.
+     */
+    private fun getLevelsTable(): ArcGISFeatureTable? {
+        return try {
+            val levelsFeatureLayer =
+                mapView.map?.operationalLayers?.firstOrNull { it.name == "levels" } as? FeatureLayer
+            levelsFeatureLayer?.featureTable as? ArcGISFeatureTable
+        } catch (e: Exception) {
+            // if levels table not found in map's operationalLayers
+            showError("Levels table not found")
+            null
+        }
+    }
```

Comment:
The long null check chains in these functions are kinda hard to follow. You could rescope it with a `let` and `?:` with a `run` block if you a message has to be shown. Also I think the `getPathwaysTable()` and `getLevelsTable()` could be made into a single function that takes the layer `name` as the parameter.

```suggestion
    private fun getFeatureTable(name: String): ArcGISFeatureTable? {
        return mapView.map?.operationalLayers?.firstOrNull { operationalLayer ->
            operationalLayer.name == name
        }?.let { layer -> 
            if (layer is FeatureLayer) {
                layer.featureTable as ArcGISFeatureTable
            } else {
                null
            }
        } ?: run {
            showError("$name table not found")
            null
        }
    }
```


Something like this should work.